File: prg4.cpp
#include<iostream.h>
#include<stdlib.h>
#include<conio.h>
#include<values.h>
#include<stdio.h>
#include<fstream.h>
#include<string.h>
fstream fp;
int n=0;
char fname[10],index[10];
class student
{
 private:
 char name[15],usn[15],branch[5];
 public:
 void pack();
 void unpack(char[]);
 void display();
 void searchbyrrn();
};
void student::pack()
{
 char buffer[100];
 cout<<"Enter the student USN, Name and Branch:";
 cin>>usn>>name>>branch;
 strcpy(buffer,usn);
 strcat(buffer,"|");
 strcat(buffer,name);
 strcat(buffer,"|");
 strcat(buffer,branch);
 index[n-1] = fp.tellg();
 fp<<buffer;
 fp<<"\n";
}
void student::unpack(char buffer[])
{
 char *t;
 t = strtok(buffer,"|");
 cout<<"\nUSN : "<<t<<endl;
 t = strtok(NULL,"|");
 cout<<"Name : "<<t<<endl;
 t = strtok(NULL,'\0');
 cout<<"Branch :"<<t<<endl;
}
void student::display()
{
 char buffer[100];
 int i,j;
 cout<<"\nFile Contents are \n";
 fp.open(fname,ios::in);
 for(i=1;i<=n;i++)
 {
 fp>>buffer;
 unpack(buffer);
 }
 fp.close();
}
void student::searchbyrrn()
{
 int rrn;
char buffer[30];
 cout << "\nEnter RRN of the record:";
 cin>>rrn;
 fp.open(fname,ios::in);
 if(rrn > n-1)
 {
 cout <<"Record with RRN "<<rrn<<" was not found..\n";
 fp.close();
 return;
 }
 fp.seekg(index[rrn],ios::beg);
 fp>>buffer;
 unpack(buffer);
 fp.close();
 return;
}
void main()
{
 int i,ch;
 student s1;
 clrscr();
 cout<<"\nEnter the Filename:";
 cin>>fname;
 fp.open(fname,ios::out|ios::trunc);
 fp.close();
 for(;;)
 {
 cout<<"\n 1.Insert \n 2.Display \n 3.Search By RRN \n 4.Exit \n";
 cout<<"Enter your Choice:";
 cin>>ch;
 switch(ch)
 {
 case 1:
 n++;
 fp.open(fname,ios::out|ios::ate);
 s1.pack();
 fp.close();
 break;
 case 2:
 s1.display();
 break;
 case 3:
 s1.searchbyrrn();
 break;
default:
 exit(0);
 }
 }
}

File: prg4/prg4.py
import sys
import os


def checkMaxLen(val: str, max: int):
    if len(val) > max:
        print("exceeded maximum limit!")
        sys.exit()

    return val


class Student:
    # private data members
    __USNMAX = 15
    __NAMMAX = 15
    __BRNMAX = 5

    def __init__(self, filename):
        self.filename = filename
        self.recordsNum = 0
        self.index = []
        self.__initializeFile()

    def __initializeFile(self):
        file = open(self.filename, "w")
        file.close()

    def pack(self):
        usn = checkMaxLen(input("usn: "), self.__USNMAX)
        name = checkMaxLen(input("name: "), self.__NAMMAX)
        branch = checkMaxLen(input("branch: "), self.__BRNMAX)

        buffer = usn + "|" + name + "|" + branch + "|"

        with open(self.filename, "a") as f:
            self.recordsNum = self.recordsNum + 1
            self.index.append(f.tell())

            f.write(f"{buffer}\n")

    def unpack(self, record):
        [usn, name, branch, *_] = record.split("|")
        print(f"USN: {usn}\nName: {name}\nBranch: {branch}\n")

    def display(self):
        if self.recordsNum == 0:
            print("No records to display!")
            return

        print("\nFile contents are: ")

        with open(self.filename, "r") as f:
            records = [record.strip() for record in f.readlines()]
            for record in records:
                self.unpack(record)

    def searchByRRN(self):
        rrn = int(input("Enter the RRN of the record: "))
        if rrn > self.recordsNum - 1:
            print(f"Record with RRN {rrn} not found!")
            return

        f = open(self.filename, "r")
        f.seek(self.index[rrn])
        self.unpack(f.readline())
        f.close()


if __name__ == "__main__":
    filename = input("Enter file name: ")
    stu = Student(filename)

    while True:
        ch = int(
            input(
                "\n1. Insert\n2. Display\n3. Search by RRN\n4. Exit\nEnter your choice: "
            )
        )

        if ch == 1:
            stu.pack()
        elif ch == 2:
            stu.display()
        elif ch == 3:
            stu.searchByRRN()
        elif ch == 4:
            break
        else:
            print("Invalid choice\n")
